text/temp/PALENICA/Larsson_Sadakane_Suffix_Array.tex

```diff
@@ -0,0 +1,113 @@
+\end{itemize}
+\end{definition}
+Grupy będziemy numerować żeby móc przeprowadzać obliczenia na poszczególnych grupach z osobna. Dla grupy $I[f \ldots g]$ numer tej grupy to $g$. Ponadto tablica $V[i] = g$ zawiera informację, że sufiks $S_{i+1}$ znajduje się w grupie o numerze $g$. Dodatkowo będziemy korzystać z tablicy $L$ która trzyma długości nieposortowanych grup oraz długości wspólnych grup posortowanych. Dla rozróżnienia, długości tych późniejszych będą trzymane jako liczba ujemna. Tablicę $L$ użyjemy w podstawowej wersji algorytmu. Udoskonalenia omówione pod koniec tego omówienia pozwolą nam obyć się bez osobnej tablicy na długości grup.
+\begin{observation}
+\label{obs2}
+Gdy I jest w \textit{h-order}, każdy sufiks w \textbf{wspólnej grupie posortowanej} jest unikalnie rozróżnialny od wszystkich innych sufiksów przy pomocy pierwszych $h$ symboli.
+\end{observation}
+
+\section*{Algorytm}
+
+Algorytm po każdej iteracji konstruuje z tablicy sufiksów w porządku \textit{h-order} tablicę w porządku \textit{2h-order} korzystając z obserwacji \ref{obs1}. Zgodnie z obserwacją \ref{obs2}, sufiksy znajdujące się we wspólnych grupach posortowanych są już na swoich miejscach. Mając to na uwadze możemy wykorzystać to w algorytmie i przy procesowaniu nieposortowanych grup sufiksów `przeskakiwać` nad wspólnymi grupami posortowanymi. Po każdej iteracji jest szansa, że pojawią się nowe grupy posortowane więc dodatkowo będziemy pilnować aby spójne sekwencje grup posortowanych były przez nas łączone we wspólne grupy posortowane.
+
+Podstawowy algorytm będzie się prezentował następująco:
+\begin{enumerate}
+    \item Umieść sufiksy $1 \ldots n+1$ w tablicy $I$. Posortuj $I$ używając $x_i$ jako klucza dla sufiksu $i$ (pierwsza litera). Ustaw $h=1$
+    \item Dla każdego sufiksu $S_{i+1}$ ustaw numer grupy do której należy w $V[i]$
+    \item Dla każdej nieposortowanej grupy lub wspólnej grupy posortowanej zapisz jej długość (lub zanegowaną długość) do tablicy $L$
+    \item Przesortuj każdą nieposortowaną grupę z osobna za pomocą \textit{ternary quick sorta} (czyli takiego co dzieli na podtablice $[< \textit{pivot}, = \textit{pivot}, > \textit{pivot}]$), używając $V[S_k + h - 1]$ jako klucza dla każdego sufiksu $S_k$ w grupie
+    \item Zaznacz pozycje podziału pomiędzy nierównymi kluczami w dla każdej przetworzonej nieposortowanej grupy na potrzeby aktualizacji tablic $V$ oraz $L$
+    \item Podwój $h$. Stwórz nowe grupy poprzez podział na zaznaczonych pozycjach (aktualizując odpowiednio $V$ i $L$)
+    \item Zakończ jeśli $I$ składa się tylko z jednej posortowanej grupy. W przeciwnym przypadku idź do kroku 4.
+\end{enumerate}
+
+W podstawowej wersji algorytmu tablice $V$ oraz $L$ aktualizujemy po wykonaniu sortowania. Możemy to wykonać np. zaznaczając pozycje podziałów w procedurze quicksort np. używając bitów znaku w tablicy $I$. Na tej podstawie aktualizujemy $V$ - ujemny wpis oznacza nową grupę - a następnie tablicę $L$.
```

Comment:
"Możemy to wykonać np. zaznaczając pozycje podziałów w procedurze quicksort np. używając bitów znaku w tablicy $I$." - still it's somewhat hard to comprehend by the reader. Maybe an example would be helpful?